# Jiaming Lin

## 1 Confusion Matrix (45pt)

#### 1. (4pt) Show the confusion matrix for M1 and M2.

M1: Yes for simolestes vorax, No for other marine dinosaurs

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 3         | 2           |
| Yes(actual)        | 1      | 4               |

M2: Yes for simolestes vorax, No for other marine dinosaurs

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 2         | 3           |
| Yes(actual)        | 0      | 5               |

#### 2. (5pt) Compute accuracy, precision, recall, and F-score for both models.

M1: A: 0.7, P: 0.6667, R: 0.8, F-score: 0.7273

M2: A: 0.7, P: 0.625, R: 1, F-score: 0.769

#### 3. (6pt) Prof Smith is going to publish her paper and wants you to give the final results tomorrow. Which of these two models, M1 or M2 will you recommend her? Explain to her.

I will recommend M2 for her. The reason is that both models have same precentage correctness in prediction and M1 has a not too significantly higher accuracy when predict it is simolestes vorax than M2, but M2 can predict every simolestes vorax correct if it really is. Besides, M2 performs better when we combine previous two comparisions together.

#### 4. (4pt) Show the confusion matrices for M3 and M4.

M3: Yes for poisoneus mushrooms, No for edible mushrooms

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 4         | 3           |
| Yes(actual)        | 0      | 3               |

M4: Yes for poisoneus mushrooms, No for edible mushrooms

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 7         | 0           |
| Yes(actual)        | 1      | 2               |

#### 5. (5pt) Compute the accuracy, precision, recall for both models.


M3: A: 0.7, P 0.5, R: 1

M4: A: 0.9, P 1, R: 0.667

#### 9. (6pt) Johanknecht wants to commission the AI system tomorrow. Which model would you recommend her to use? Explain your reasoning!

I will recommend her M3 model, because although its accuracy is lower than M4, but it can predict every poisoneus mushroom correct if it is, which is more important than accuracy because we don't want to see people eat poisoneus mushrooms.

#### 7. (4pt) Show the confusion matrices for M5 and M6.

M5: Yes for guilty, No for innocent

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 3         | 2           |
| Yes(actual)        | 1      | 4               |

M6: Yes for guilty, No for innocent

|            | No(predict) | Yes(predict)               |
| :---        |    :----:   |          ---: |
| No(actual)        | 5         | 0           |
| Yes(actual)        | 3      | 2               |

#### 8. (5pt) Compute the accuracy, precision, recall for both models.

M5: A: 0.7, P: 0.667, R: 0.8
                
M6: A: 0.7, P: 1, R: 0.4

#### 9. (6pt) Johanknecht wants to commission the AI system tomorrow. Which model would you recommend her to use? Explain your reasoning!


I will recommend to her M5, because both modles have same accuracy, but M5 predicts much better if the defendants is guilty, and we don't want to see those guilty defendants still active in the society

## 2 Cross-validate to the best model (55pt)

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score
from sklearn.model_selection import cross_val_score, train_test_split
import warnings
warnings.filterwarnings('ignore')

#### 1. (2pt) Load the data. Ensure you understand the variables well enough (but no need to learn their biological meaning). We recommend to consult the uploaded doc file.

#### Remove the id variable

In [2]:
wc = pd.read_csv("wdbc.csv.bz2")
wc = wc.drop(["id"], axis=1)
wc.head()

diagnosis  radius.mean  texture.mean  perimeter.mean  area.mean  \
0         M        17.99         10.38          122.80     1001.0   
1         M        20.57         17.77          132.90     1326.0   
2         M        19.69         21.25          130.00     1203.0   
3         M        11.42         20.38           77.58      386.1   
4         M        20.29         14.34          135.10     1297.0   

   smoothness.mean  compactness.mean  concavity.mean  concpoints.mean  \
0          0.11840           0.27760          0.3001          0.14710   
1          0.08474           0.07864          0.0869          0.07017   
2          0.10960           0.15990          0.1974          0.12790   
3          0.14250           0.28390          0.2414          0.10520   
4          0.10030           0.13280          0.1980          0.10430   

   symmetry.mean  ...  radius.worst  texture.worst  perimeter.worst  \
0         0.2419  ...         25.38          17.33           184.60   
1         0.1812  ...         24.99          23.41           158.80   
2         0.2069  ...         23.57          25.53           152.50   
3         0.2597  ...         14.91          26.50            98.87   
4         0.1809  ...         22.54          16.67           152.20   

   area.worst  smoothness.worst  compactness.worst  concavity.worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concpoints.worst  symmetry.worst  fracdim.worst  
0            0.2654          0.4601        0.11890  
1            0.1860          0.2750        0.08902  
2            0.2430          0.3613        0.08758  
3            0.2575          0.6638        0.17300  
4            0.1625          0.2364        0.07678  

[5 rows x 31 columns]

#### 2. (2pt) Create your outcome vector y. While LogisticRegression can easily handle string labels B and M, the f1_score cannot. So we recommend you to convert the diagnosis into a numeric 1/0 label.


In [3]:
wc["y"] =  wc["diagnosis"]
wc["y"].loc[wc["y"] == "M"] = "1"
wc["y"].loc[wc["y"] == "B"] = "0"
y = wc["y"]

#### 3. (4pt) Create the design matrix X that contains three arbitrary columns from among your features. I recommend to use .iloc, e.g. X = data.iloc[:,[1,5,24]].values to make a design matrix from these three columns.

In [4]:
X = wc.iloc[:,[1,2,3]]

#### 4. (6pt) Fit a logistic regression model predicting y using this design matrix X.

In [5]:
m = LogisticRegression()
_ = m.fit(X, y)

#### 5. (9pt) Compute the predicted outcomes, and display the confusion matrix and F-score.

In [6]:
yhat = m.predict(X)
confusion_matrix(y, yhat)

array([[337,  20],
       [ 32, 180]], dtype=int64)

In [7]:
y = [int(numeric_string) for numeric_string in y]
yhat = [int(numeric_string) for numeric_string in yhat]
f1_score(y, yhat)
# The F-score is 0.8737864077669903

0.8737864077669903

#### 6. (9pt) What do you think, what is the most appropriate measure here? Why is F-score a good way to test the models?


I will say recall, because is important here to predict right if a person has cancer. F-score is always a good way because it combines both recall and precision, so it is good method to see how models work overall.

#### 7. (9pt) Cross-validate the model goodness using F-score as the outcome. (use 5-fold or more) Now it is time to check all possible 3-feature models. You can do three nested loops that run over the columns, create X out of those columns, and repeat the above

In [8]:
m = LogisticRegression()
_ = m.fit(X, y)
cv = cross_val_score(m, X, y, scoring="f1", cv=5)
print(cv.mean())
# The F-score is 0.8733951787980809

0.8733951787980809


#### 8. (9pt) Loop over all 3-feature combinations and:

#### (a) at the corresponding logistic regression model
#### (b) cross-validate the F-score

In [9]:
m = LogisticRegression()
comb = np.array([])
f1s = np.array([])

for i in range(1, 29):
    for j in range(i+1,30):
        for k in range(j+1,31):
            X = wc.iloc[:,[i,j,k]]
            _ = m.fit(X, y)
            cv = cross_val_score(m, X, y, scoring="f1", cv=5)
            comb = np.append(comb, str(i)+","+str(j)+","+str(k))
            f1s = np.append(f1s, cv.mean())            

In [11]:
index = np.where(f1s == np.amax(f1s))
print(index)
print(f1s[20])
print(comb[20])

(array([20], dtype=int64),)
0.9367686074693806
1,2,23


#### 9. (5pt) Finally, tell what is your best model and its F-score and accuracy. Which three features did you include?

In [13]:
X = wc.iloc[:,[1,2,23]]
cv = cross_val_score(m, X, y, scoring="f1", cv=5)
print("f1:", cv.mean())
cv = cross_val_score(m, X, y, scoring="accuracy", cv=5)
print("Accuracy:", cv.mean())

coli = wc.columns[1]
colj = wc.columns[2]
colk = wc.columns[23]
print(coli,colj,colk)

f1: 0.9367686074693806
Accuracy: 0.9543238627542306
radius.mean texture.mean perimeter.worst


The best model is column 1(radius.mean), 2(texture.mean) and 23(perimeter.worst), F-score for this model is 0.9368 and accuracy is 0.9543

#### How much time did you spend on this PS?

I spend 4 hours